src/Compiler/Microsoft.AspNetCore.Mvc.Razor.Extensions/src/ViewComponentTagHelperDescriptorProvider.cs

```diff
@@ -37,43 +38,39 @@ public void Execute(TagHelperDescriptorProviderContext context)
             return;
         }
 
-        var types = new List<INamedTypeSymbol>();
-        var visitor = new ViewComponentTypeVisitor(vcAttribute, nonVCAttribute, types);
+        var factory = new ViewComponentTagHelperDescriptorFactory(compilation);
+        var collector = new Collector(compilation, factory, vcAttribute, nonVCAttribute);
 
-        var targetSymbol = context.Items.GetTargetSymbol();
-        if (targetSymbol is not null)
-        {
-            visitor.Visit(targetSymbol);
-        }
-        else
-        {
-            visitor.Visit(compilation.Assembly.GlobalNamespace);
-            foreach (var reference in compilation.References)
-            {
-                if (compilation.GetAssemblyOrModuleSymbol(reference) is IAssemblySymbol assembly)
-                {
-                    if (IsTagHelperAssembly(assembly))
-                    {
-                        visitor.Visit(assembly.GlobalNamespace);
-                    }
-                }
-            }
-        }
+        collector.Collect(context);
+    }
 
-        var factory = new ViewComponentTagHelperDescriptorFactory(compilation);
-        for (var i = 0; i < types.Count; i++)
+    private class Collector(
+        Compilation compilation,
+        ViewComponentTagHelperDescriptorFactory factory,
+        INamedTypeSymbol vcAttribute,
+        INamedTypeSymbol nonVCAttribute)
+        : TagHelperCollector<ViewComponentTagHelperDescriptorProvider>(compilation, targetSymbol: null)
+    {
+        private readonly ViewComponentTagHelperDescriptorFactory _factory = factory;
+        private readonly INamedTypeSymbol _vcAttribute = vcAttribute;
+        private readonly INamedTypeSymbol _nonVCAttribute = nonVCAttribute;
```

Comment:
I wonder why you chose not to simply use the primary constructor parameters and instead stored them to fields explicitly? Feels redundant.

Reply:
Honestly, because it's controversial. Some folks have expressed that they prefer to see all of the instance data of a type as fields at the top of this type, and this approach is a happy medium to me.